Code/src/main/java/nl/utwente/group10/ui/components/blocks/function/ArgumentSpace.java

```diff
@@ -0,0 +1,352 @@
+            //This is needed for when the knot does not have a position yet (initializing).
+            return getKnotIndex();
+        }
+        
+        double bowtieIndex = 0;
+        for (int i = 0; i < leftArguments.size(); i++) {
+            Region argument = leftArguments.get(i);
+            double min = argument.getLayoutX();
+            double max = min + argument.getWidth();
+            if (getKnotPos() > max) {
+                bowtieIndex++;
+            } else {
+                bowtieIndex += Math.max(Math.min((getKnotPos() - min) / (max - min), 1), 0);
+                break;
+            }
+        }
+        return bowtieIndex;
+    }
+
+    /**
+     * This repositions the knot to a position based on the current knot index,
+     * regardless of the knot's current position.
+     * 
+     * Changes in Label length after snapping are accounted for by using
+     * property bindings.
+     */
+    public void snapToKnotIndex() {
+        int kti = getKnotIndex();
+        if (kti > 0 && kti <= leftArguments.size()) {
+            Region arg = leftArguments.get(kti-1); // First argument left of the knot.
+
+            // Binds the knot to the first argument on the left, so that if
+            // there is a length change to the left of the knot (because of a
+            // Label's text change), the knot moves accordingly.
+            knot.layoutXProperty().bind(arg.layoutXProperty().add(arg.widthProperty()).add(knot.radiusProperty()).add(H_GAP));
+        }else if (kti == 0) { // There is no argument left of the knot.
+            knot.setLayoutX(knot.getRadius());
+        }
+        
+        invalidateKnotPosition();
+    }
+    
+    /**
+     * Method to indicate that the position of the knot might have changed.
+     * 
+     * It repositions the arguments based on the current position of the knot.
+     */
+    public void invalidateKnotPosition() {
+        double bti = determineKnotIndex();
+        for (int i = 0; i < leftArguments.size(); i++) {
+            Region argument = leftArguments.get(i);
+            // Get only the percentage for this argument, clamped between 0 and 1, inverted.
+            double percentage = Math.min(Math.max(bti - i,  0), 1);            
+            percentage = Math.pow(percentage, 0.4); //Better looking curve
+            
+            //Updates the argument based on the percentage
+            argument.setTranslateX((H_GAP +knot.getRadius() * 2) * (1 - percentage));
+            argument.setOpacity(percentage);
+            argument.setVisible(percentage > 0);
+        }
+    }
+    
+    /**
+     * Method to indicate that the content in the argument Labels are possibly outdated.
+     */
+    public void invalidateArgumentContent() {
+        invalidateOutputContent();
+        invalidateInputContent();
+    }
+    
+    /**
+     * Method to indicate that the content in the output argument Label is possibly outdated.
+     */
+    public void invalidateOutputContent() {
```

Comment:
Currently not possible, but these methods have no side effects, so no real problems arise from this.